src/plugins/data/common/search/search_source/search_source.ts

```diff
@@ -507,6 +516,76 @@ export class SearchSource {
     return searchRequest;
   }
 
+  private flattenWithFieldsApi(searchRequest: Record<string, any>) {
+    const { getConfig } = this.dependencies;
+
+    searchRequest.body = searchRequest.body || {};
+    const { body, index, query, filters, highlightAll } = searchRequest;
+    searchRequest.indexType = this.getIndexType(index);
+
+    const fieldListProvided = !!body.fields;
+
+    // get some special field types from the index pattern
+    const { storedFields, scriptFields, docvalueFields } = index
+      ? index.getComputedFields()
+      : {
+          storedFields: ['*'],
+          scriptFields: {},
+          docvalueFields: [],
+        };
+
+    // add docvalues to fields if no fields are explicitly provided
+    body.fields = fieldListProvided ? body.fields : docvalueFields;
+
+    // apply source filters from index pattern if specified by the user
+    if (index) {
+      const sourceFilters = index.getSourceFiltering();
+      if (!body.hasOwnProperty('_source')) {
+        body._source = sourceFilters;
+      }
+      const filter = fieldWildcardFilter(
+        sourceFilters.excludes,
+        getConfig(UI_SETTINGS.META_FIELDS)
+      );
+      body.fields = body.fields.filter((fld: any) =>
+        filter(typeof fld === 'string' ? fld : fld.field)
+      );
+    }
+
+    // add scripted fields
+    body.script_fields = pick(Object.assign(body.script_fields || {}, scriptFields), body.fields);
```

Comment:
seems we handle script_fields exactly the same as before (just the code is cleaned up) ?

Reply:
same goes for stored_fields and docvalue_fields